### PR TITLE
Improve performance

### PR DESCRIPTION
**Pull request by pverscha**
This PR drastically improves the script's performance by implementing two optimizations:

1) Precompute the information content of the most informative ancestor. If the script loads for the first time, it checks if the precomputed information is available. If not, it precomputes the data, stores it as a JSON-file and loads in memory. Invocations of the script afterwards automatically use the precomputed data.

2) Computing the deepest common ancestor is cached. Once this value has been computed for a set of inputs, it will be reused.

I also split the files some more to be more concise and renamed a few input files for consistency.

### CODE REVIEW

File: megago/megago.py [hunk 1]
```diff
@@ -85,172 +89,15 @@ def read_input(in_file, sep=",", go_sep=";"):
...
-def parse_args():
-    """Parse command line arguments.
-    Returns Options object with command line argument values as attributes.
-    Will exit the program on a command line error.
+    Options object with command line argument values as attributes.
     """
-
-    parser = argparse.ArgumentParser(description=__doc__)
+    description = 'Calculate semantic distance for sets of Gene Ontology terms'
+    parser = argparse.ArgumentParser(description=description)
```
Comment (rababerladuseladim) [inline megago/megago.py:100]: Why not use the doc string of the module? I introduced the usage of the doc string based on [this](https://realpython.com/documenting-python-code/#documenting-your-python-code-base-using-docstrings)

Comment (rababerladuseladim) [reply megago/megago.py:100]: Please don't get me wrong, I am not criticizing your changes but rather would like to know your rationale.

File: megago/megago.py [hunk 2]
```diff
@@ -35,15 +33,21 @@
...
+    Returns
+    -------
+    True if the string is indeed a valid identifier, False otherwise.
+    """
     regex = re.compile(r"^go:\d{7}$", re.IGNORECASE)
-    if regex.match(string):
-        return True
-    else:
-        return False
+    return regex.match(candidate)
```
Comment (rababerladuseladim) [inline megago/megago.py:50]:
```suggestion
    return bool(regex.match(candidate))
```

The regex.match does not return a bool but a match-object or None.

File: megago/megago.py [hunk 3]
```diff
@@ -35,15 +33,21 @@
...
+    satisfies the formatting of a GO-term identifier, not if it actually exists in the ontology!
 
-def is_go_term(string):
+    Parameters
+    ----------
+    candidate: A string for which we should verify if it's a valid GO-identifier.
+
+    Returns
+    -------
+    True if the string is indeed a valid identifier, False otherwise.
```
Comment (rababerladuseladim) [inline megago/megago.py:47]:
```suggestion
    Parameters
    ----------
    candidate : str
        A string for which we should verify if it's a valid GO-identifier.

    Returns
    -------
    bool
```
Numpy-style 😎
Do you maybe know an easy way for validating code style conformitty? I have heard of such things but never used them myself. Would be a cool feature to add to our CI.

File: megago/metrics.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+       gene ontology ID that the relative frequency should be calculated for
+   term_counts : dict
+       dictionary: key: GO terms, values: number of occurrences of GO term and its children in body of evidence
+   go_dag : GODag object
+       GODag object from the goatools package
+
+   Returns
+   -------
+   float
+   """
```
Comment (rababerladuseladim) [inline megago/metrics.py:29]: indent is of by one, I think

File: setup.py [hunk 5]
```diff
@@ -20,9 +20,12 @@
...
-                             "resource_data/go_freq_uniprot.json"]},
+    package_data={'megago': [
+        "resources/associations-uniprot-sp-20200116.tab",
+        "resources/go-basic.obo",
+        "resources/frequency_counts_uniprot.json",
+        "resources/highest_ic_uniprot.json"
+    ]},
     entry_points={
         'console_scripts': ['megago = megago.megago:main']
     },
```
Comment (rababerladuseladim) [inline setup.py:31]: add `progress` requirement to install_requires

File: timing/timings.csv [hunk 6]
```diff
@@ -1,56 +1,73 @@
...
+3150,3150,49.193811893463135
+3200,3200,50.16846799850464
+3250,3250,53.15678524971008
+3300,3300,54.0193407535553
+3350,3350,54.78669500350952
+3400,3400,56.85016393661499
+3450,3450,56.19050598144531
+3500,3500,56.335593938827515
+3550,3550,59.445597887039185
+3600,3600,58.73915886878967
```
Comment (rababerladuseladim) [inline timing/timings.csv:73]: AMAZING!!!

File: megago/megago.py (hunk 1 above)
Comment (pverscha) [reply megago/megago.py:100]: I must have made a mistake while resolving a merge conflict. Yours is the right way to go.

File: megago/megago.py (hunk 3 above)
Comment (pverscha) [reply megago/megago.py:47]: We could use the Python linter for that. It's a tool that checks if the code adheres to specific rules that we set up. I'll take a look to implement a new Github action with these checks (we could maybe use this: https://github.com/marketplace/actions/python-lint)

File: timing/timings.1.csv [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+2400,2400,169.11304092407227
+2450,2450,178.14109706878662
+2500,2500,182.99004101753235
+2550,2550,192.63860201835632
+2600,2600,202.99907779693604
+2650,2650,209.94929814338684
+2700,2700,216.70873308181763
+2750,2750,220.71831011772156
+2800,2800,233.41998600959778
+2850,2850,235.86974000930786
```
Comment (rababerladuseladim) [inline timing/timings.1.csv:58]: What's the difference between timings and timings.1?

Comment (pverscha) [reply timing/timings.1.csv:58]: `Timings.1` contains measurements after I applied the first optimization, `timings` contains the measurements with both the first and second optimization enabled. I think I'll just delete `timings.1` as we still have these files in a Google spreadsheet.

Comment (rababerladuseladim) [reply timing/timings.1.csv]: Could you link the google spreadsheet to the code somehow? Prabably an optimization issue is the most sensible place.

File: megago/megago.py [hunk 8]
```diff
@@ -303,79 +151,37 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+def run_comparison(in_file):
+    """ Compute the pairwise similarity values for all rows from the given file.
+    Params
+    ------
+    in_file: The file for wich row-wise similarities should be computed.
+
+    Returns
+    -------
+    A string that represents a "CSV"-file with a similarity value per row.
+    """
```
Comment (rababerladuseladim) [inline megago/megago.py:163]:
```suggestion
    """ Compute the pairwise similarity values for all rows from the given file.
    
    Parameters
    ------
    in_file : file_handle / path_to_file
        The file for which row-wise similarities should be computed.

    Returns
    -------
    str
        A string that represents a "CSV"-file with a similarity value per row.
    """
```

I am not sure about the type of the `in_file` parameter.

File: .gitignore [hunk 9]
```diff
@@ -1,23 +1,156 @@
...
+# pytype static type analyzer
+.pytype/
+
+# Cython debug symbols
+cython_debug/
+
+.DS_Store
+
 
 .ipynb_checkpoints/similarty_metrics-checkpoint.ipynb
```
Comment (rababerladuseladim) [inline .gitignore:143]:
```suggestion

```

File: timing/timings.1.csv (hunk 7 above)
Comment (pverscha) [reply timing/timings.1.csv]: Good idea, I added the link here: https://github.com/MEGA-GO/Mega-GO/issues/13#issuecomment-627795374

File: .gitignore [hunk 10]
```diff
@@ -1,23 +1,156 @@
...
-.idea/Mega-Go.iml
-.idea/misc.xml
-.idea/modules.xml
-.idea/vcs.xml
-.idea/workspace.xml
+.idea/**/*
 *.iml
 aaa_lin.png
-venv/
 megago/.ipynb_checkpoints/similarty_metrics-checkpoint.ipynb
```
Comment (rababerladuseladim) [inline .gitignore:155]:
```suggestion

```

**Pull request merged**